### PR TITLE
Use reference type from the pmap and fix warnings

### PR DESCRIPTION
**Pull request by sloriot**
At the same time I updated examples that were using an undocumented API

### CODE REVIEW

**Comment by sloriot** [comment]
Actually I found out that the code was insecure as it was modifying the mesh while iterating over descriptors. I fixed that by copying first the descriptors into vectors. This enabled the removal of the two implementations for DQQ + clean up of the manual (no more the garbage free condition for Surface_mesh).

Here are the runtime I have for 3 iterators over the elephant:
```
DooSabin_subdivision-new  (0.250502 s)
DooSabin_subdivision-4.11 (0.201872 s)
DooSabin_subdivision_poly_index-new (0.212643 s)
DooSabin_subdivision_poly_index-4.11  (0.371118 s)
DooSabin_subdivision_sm-new  (0.076591 s)
DooSabin_subdivision_sm-4.11  (0.179114 s)
Sqrt3_subdivision-new  (0.022829 s)
Sqrt3_subdivision-4.11  (0.030424 s)
Customized_subdivision-new  (0.028928 s)
Customized_subdivision-4.11  (0.036304 s)
Loop_subdivision-new  (0.029892 s)
Loop_subdivision-4.11 (0.036008 s)
CatmullClark_subdivision-new (0.063742 s)
CatmullClark_subdivision-4.11  (0.061143 s)
```

File: Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt [hunk 1]
```diff
@@ -127,12 +127,12 @@ There is only one line deserving a detailed explanation:
 
 \code{.cpp}
 
-Subdivision_method_3::CatmullClark_subdivision(pmesh, d);
+Subdivision_method_3::CatmullClark_subdivision(pmesh, params::number_of_iterations(d));
```
Comment (MaelRL) [inline Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt]: `CGAL::parameters` (same for others)

File: Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt [hunk 2]
```diff
@@ -188,7 +182,7 @@ three components of the Catmull-Clark subdivision method.
 \code{.cpp}
 
 template <class PolygonMesh, class Mask>
-void PQQ(PolygonMesh& p, Mask mask, int depth)
+void PQQ(PolygonMesh& p, Mask mask, NamedParameters np)
```
Comment (MaelRL) [inline Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt]: `NamedParameters` in template

File: Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt [hunk 3]
```diff
@@ -203,8 +197,7 @@ new points by cooperating with the `mask`.
 To implement Catmull-Clark subdivision,
 `Mask`, the <I>geometry policy</I>, has to realize the
 geometry masks of Catmull-Clark subdivision.
-The parameter `depth` specifies the iterations of the refinement
-on the control mesh.
+The number of iterations as well as the vertex point map can be specified using the named parameter `np`.
```
Comment (MaelRL) [inline Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt]: I'll fix that later once https://github.com/CGAL/cgal/pull/2649 is through.

File: Subdivision_method_3/include/CGAL/Subdivision_method_3/subdivision_masks_3.h [hunk 4]
```diff
@@ -116,8 +118,8 @@ class Linear_mask_3 : public PQQ_stencil_3<PolygonMesh, VertexPointMap> {
   }
 
   void edge_node(halfedge_descriptor edge, Point& pt) {
-    const Point& p1 = get(this->vpmap, target(edge, *(this->pmesh)));
-    const Point& p2 = get(this->vpmap, source(edge, *(this->pmesh)));
+    const Point_ref p1 = get(this->vpmap, target(edge, *(this->pmesh)));
+    const Point_ref p2 = get(this->vpmap, source(edge, *(this->pmesh)));
```
Comment (MaelRL) [inline Subdivision_method_3/include/CGAL/Subdivision_method_3/subdivision_masks_3.h]: There are others that require the same changes in this file and also in `test_subdivision_method_3.cpp` and `Customized_subdivision_3.cpp`

File: Subdivision_method_3/include/CGAL/Subdivision_method_3/internal/subdivision_hosts_impl_3.h [hunk 5]
```diff
@@ -312,13 +316,14 @@ void DQQ_1step_impl(Poly& p, VertexPointMap vpm, Mask mask, CGAL::Tag_false) {
...
-  for (size_t i = 0; i < num_v; ++i) {
+  typename std::vector<vertex_descriptor>::iterator vitr=p_vertices.begin();
+  for (std::size_t i = 0; i < num_v; ++i) {
     vertex_descriptor vh = *vitr;
     ++vitr;
 
     Halfedge_around_target_circulator<Poly> vcir(vh,p);
-    size_t vn = degree(vh,p);
-    for (size_t j = 0; j < vn; ++j) {
+    typename boost::graph_traits<Poly>::degree_size_type vn = degree(vh,p);
```
Comment (MaelRL) [inline Subdivision_method_3/include/CGAL/Subdivision_method_3/internal/subdivision_hosts_impl_3.h]: Same change later in the file (l 580)

File: Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt (hunk 1 above)
Comment (sloriot) [reply Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt]: We are in 4.11, it should be `CGAL::Polygon_mesh_processing::parameters` but I'm using the alias from the example.

File: Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt (hunk 3 above)
Comment (sloriot) [reply Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt]: What do you mean?

Comment (MaelRL) [reply Subdivision_method_3/doc/Subdivision_method_3/Subdivision_method_3.txt]: I mean detailing a bit what named parameters are and linking to the NamedParameters.txt of the package.

File: Subdivision_method_3/include/CGAL/Subdivision_method_3/internal/subdivision_hosts_impl_3.h (hunk 5 above)
Comment (sloriot) [reply Subdivision_method_3/include/CGAL/Subdivision_method_3/internal/subdivision_hosts_impl_3.h]: Fixed

**Comment by sloriot** [comment]
Travis error is due to apt-get error: `apt-get install failed`

**Comment by lrineau** [comment]
I have restarted the problematic Travis job.

**Comment by lrineau** [comment]
There is [a new warning in Polyhedron_Demo][*]:
```
C:\CGAL\test\CGAL-4.12-Ic-117\include\CGAL/Subdivision_method_3/internal/subdivision_hosts_impl_3.h(437) : warning C4267: 'argument' : conversion from 'size_t' to 'unsigned int', possible loss of data
        C:\CGAL\test\CGAL-4.12-Ic-117\include\CGAL/Subdivision_method_3/subdivision_hosts_3.h(203) : see reference to function template instantiation 'void CGAL::Subdivision_method_3::internal::Sqrt3_1step<PolygonMesh,CGAL::Surface_mesh<Point_3>::Property_map<CGAL::Surface_mesh<Point_3>::Vertex_index,CGAL::Point_3<Kernel_>>,Mask>(Poly &,VertexPointMap,Mask,const bool)' being compiled
        with
        [
            PolygonMesh=Scene_surface_mesh_item::Face_graph
,            Kernel_=CGAL::Epick
,            Mask=CGAL::Sqrt3_mask_3<Scene_surface_mesh_item::Face_graph,CGAL::Surface_mesh<Point_3>::Property_map<CGAL::Surface_mesh<Point_3>::Vertex_index,CGAL::Point_3<CGAL::Epick>>>
,            Poly=Scene_surface_mesh_item::Face_graph
,            VertexPointMap=CGAL::Surface_mesh<Point_3>::Property_map<CGAL::Surface_mesh<Point_3>::Vertex_index,CGAL::Point_3<CGAL::Epick>>
        ]
        C:\CGAL\test\CGAL-4.12-Ic-117\include\CGAL/Subdivision_method_3/subdivision_methods_3.h(237) : see reference to function template instantiation 'void CGAL::Subdivision_method_3::Sqrt3<PolygonMesh,CGAL::Sqrt3_mask_3<PolygonMesh,CGAL::Surface_mesh<Point_3>::Property_map<CGAL::Surface_mesh<Point_3>::Vertex_index,CGAL::Point_3<Kernel_>>>>(PolygonMesh &,Mask,int)' being compiled
        with
        [
            PolygonMesh=Scene_surface_mesh_item::Face_graph
,            Kernel_=CGAL::Epick
,            Mask=CGAL::Sqrt3_mask_3<Scene_surface_mesh_item::Face_graph,CGAL::Surface_mesh<Point_3>::Property_map<CGAL::Surface_mesh<Point_3>::Vertex_index,CGAL::Point_3<CGAL::Epick>>>
        ]
        C:\CGAL\test\CGAL-4.12-Ic-117\cmake\platforms\x64_Cygwin-Windows10_MSVC2013-Debug-64bits\test\Polyhedron_Demo\Plugins\PCA\Basic_generator_plugin.cpp(566) : see reference to function template instantiation 'void CGAL::Subdivision_method_3::Sqrt3_subdivision<Scene_surface_mesh_item::Face_graph>(PolygonMesh &,int)' being compiled
        with
        [
            PolygonMesh=Scene_surface_mesh_item::Face_graph
        ]
        C:\CGAL\test\CGAL-4.12-Ic-117\cmake\platforms\x64_Cygwin-Windows10_MSVC2013-Debug-64bits\test\Polyhedron_Demo\Plugins\PCA\Basic_generator_plugin.cpp(302) : see reference to function template instantiation 'void Basic_generator_plugin::generateSphere<Scene_surface_mesh_item>(void)' being compiled
```

[*]: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.12-Ic-117/Polyhedron_Demo/TestReport_afabri_x64_Cygwin-Windows10_MSVC2013-Debug-64bits.gz

**Pull request merged**